### ISSUE TITLE
Can not override methods from parent class

### ISSUE THREAD

**Issue by hoanbui29** [seblyng/roslyn.nvim#72]
When I try to override the methods from parent class, nothing happened.
The methods are still showed on cmp, but I can't generate code for that. Pls help!
![CleanShot 2024-11-02 at 18 45 55@2x](https://github.com/user-attachments/assets/fc8398b0-9438-4427-83f5-8fdfa01efc24)
![CleanShot 2024-11-02 at 18 47 04@2x](https://github.com/user-attachments/assets/9f236390-50d7-4b4e-b0a2-dda35bb05ada)

**Comment by seblj** [comment]
I don't think there is much for me to do for this: https://github.com/seblj/roslyn.nvim/issues/7

I can leave this open a little bit to see if I get some time to investigate later. I cannot guarantee anything though as I am pretty sure this is just the server not following the spec (as it unfortunately doesn't do for several things)

**Comment by hoanbui29** [comment]
> I don't think there is much for me to do for this: #7
> 
> I can leave this open a little bit to see if I get some time to investigate later. I cannot guarantee anything though as I am pretty sure this is just the server not following the spec (as it unfortunately doesn't do for several things)

Thank you very much!

**Comment by drzbida** [comment]
> I don't think there is much for me to do for this: [#7](https://github.com/seblyng/roslyn.nvim/issues/7)
> 
> I can leave this open a little bit to see if I get some time to investigate later. I cannot guarantee anything though as I am pretty sure this is just the server not following the spec (as it unfortunately doesn't do for several things)

Could you check https://github.com/dotnet/roslyn/issues/76141#issuecomment-2648940065 when you have time? I've tried implementing the command in lsp_commands similar to how it's done for "roslyn.client.nestedCodeAction", following the linked vscode implementation, but I still get the exception mentioned in the issue so it doesn't get to execute. Not sure if something else is required so that the command doesn't get sent to the server (if that's why the exception is happening)

**Comment by seblyng** [comment]
Yeah, from initial testing I am also getting the same thing, and the handler is not even being called🤷‍♂️

**Comment by drzbida** [comment]
Seems like a neovim issue.

After looking through neovim 0.10.X code it seems that it's impossible for this custom command to trigger as `workspace/executeCommand` is only manually handled when triggered via code actions or code lens.

In latest neovim nightly it seems that some custom handling for commands was done in lsp -> completion.lua, but I'm still trying to figure why it still doesn't work.

**Comment by drzbida** [comment]
It seems the issue was closed only with the mentioned problem for the native completion implementation, but the core issue was not adressed. I was thinking that LSP requests for executeCommand should first check if there is a client command no matter what, but they didn't adress this part :) So currently all completion plugins need to do it on their own.

I'm also using nvim-cmp, but I'm thinking of migrating to blink as nvim-cmp seems to be kinda "unmaintained" for quite a while now.

For blink.cmp I see that this was added recently

https://github.com/Saghen/blink.cmp/pull/1130

But it's missing the part where it checks for the custom commands. I will most likely open a PR there after I migrate my config if nobody does it by then.

**Comment by seblyng** [comment]
Okay yeah thanks! I have also been thinking about moving to blink.cmp lately, but have wanted for it to get more mature first.

Anyways, thanks a lot! If you wouldn't mind, could you link this issue or just send a message here or something if you some day open a PR about that in blink? If I switch, I might also look into this, but don't know yet if/when I might do that

**Comment by drzbida** [comment]
https://github.com/Saghen/blink.cmp/pull/1255 Got merged. According to https://github.com/Saghen/blink.cmp/pull/1203 v0.12 should be the last release with breaking changes until 1.0

Caveats:
- The LSP server is sometimes very slow on the very first `completionItem` resolve that contains a command. It's possible and I hope it's just me, but the result is that sometimes the **first** override completion will not work. Video attachment here:
https://github.com/Saghen/blink.cmp/issues/1219#issuecomment-2660951930. What esentially happens is that when the completion is accepted, it has a timeout in which it can respond to the resolve, otherwise the edit is applied with the currently known information (so command is not available)

I've found that increasing blink's [resolve_timeout_ms](https://cmp.saghen.dev/configuration/reference.html#completion-accept) to 1000 ms in the newest version always works for me. 

- If you are using blink's autobrackets feature or possibly any other autobrackets functionality, you will get extra brackets after the override completion with the current implementation. Modifying the document with a command is actually a hack in LSP specs 3.17, so we must fix it here. 

```csharp
public override void _Dro // accept _DropData(Vector2 atPosition, Variant data)
```

```csharp
public override void _DropData(Vector2 atPosition, Variant data)
{
    base._DropData(atPosition, data);
}() // extra brackets here
```

- If you happen to be typing & accepting fast, *sometimes* the LSP returns broken line endings. I don't understand why this happens but I don't really think it's worth investigating
```
Notification(int what)\r\n    {\r\n        base._Notification(what);\r\n    }\r\n\r\n\r
```

Which would result in neovim in something like

```lua
public override void Notification(int what)
{
    base._Notification(what);
}

^M
```

I'll open a draft PR with fixes for the last two issues and it would be good if someone tests it more extensively (especially on something else than Windows)

**Comment by seblyng** [comment]
Thanks a lot🙌 I will see if I can setup blink.cmp again and test this out a bit.

If anyone else sees this, I hope they can also test it if interested😅

**Comment by molostovvs** [comment]
It works for me on linux.
There is indeed a problem with brackets, as described above.
Also the indexers from the autocomplete list started working:

```cs
List<int> list = [];
x. // select this[] completion item from completion list
x[] // selection result
```

**Comment by molostovvs** [comment]
Initially I wrote about the problem with parentheses without using the code from PR. 
Now I have tested the PR - parentheses are still inserted.

`override` -> `public override string ToString() => base.ToStri<cursor is here>ng();()`

**Comment by drzbida** [comment]
Thanks, I didn't have expression bodied methods enabled and forgot about this case. I've updated the PR.

There cursor will be here though

`public override string ToString() => base.ToString()<C>;`

I guess it makes the most sense for the cursor the be here

`public override string ToString() => <C>base.ToString();`

But depending on what this complex edit is used for, it might not be feasible to do without treesitter. If it's used only for override completions it should be easy to do without.

**Comment by drzbida** [comment]
I've found another problem with the autobrackets that somehow i can only trigger with ToString

`public override str` -> accept ToString

`tr` are for some reason replaced with `()` before the command executed

and the result becomes

```csharp
public override s()ing ToString()
{
    return base.ToString();
}
```

this is not solvable, it's most likely a bug in blink

**Comment by seblyng** [comment]
Interesting🤔 will try to look at the PR today or tomorrow.

> I guess it makes the most sense for the cursor the be here

I actually think it makes sense to have it as the first example you had🤷‍♂️

**Comment by seblyng** [comment]
Closing this as the handler I can implement is implemented here. The rest is up to potential completion engines to do it correctly. If it doesn't work correctly with builtin completion or blink (which I know support this), then another issue can be opened
